src/Model/Scope/CompoundCondition.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+declare(strict_types=1);
+
+namespace atk4\data\Model\Scope;
+
+use atk4\core\ContainerTrait;
+use atk4\data\Exception;
+
+/**
+ * @property AbstractCondition[] $elements
+ */
+class CompoundCondition extends AbstractCondition
+{
+    use ContainerTrait;
+
+    // junction definitions
+    public const OR = 'OR';
+    public const AND = 'AND';
+
+    /**
+     * Junction to use in case more than one element.
+     *
+     * @var self::AND|self::OR
+     */
+    protected $junction = self::AND;
+
+    /**
+     * Create a CompoundCondition from array of condition objects or condition arrays.
+     *
+     * @param AbstractCondition[]|array[] $nestedConditions
+     */
+    public function __construct(array $nestedConditions = [], string $junction = self::AND)
+    {
+        if (!in_array($junction, [self::OR, self::AND], true)) {
+            throw new Exception($junction . ' is not a valid CompondCondition junction');
+        }
+
+        $this->junction = $junction;
+
+        foreach ($nestedConditions as $nestedCondition) {
+            $nestedCondition = is_string($nestedCondition) ? new BasicCondition($nestedCondition) : $nestedCondition;
+
+            if (is_array($nestedCondition)) {
+                // array of OR nested conditions
+                if (count($nestedCondition) === 1 && isset($nestedCondition[0]) && is_array($nestedCondition[0])) {
+                    $nestedCondition = new self($nestedCondition[0], self::OR);
+                } else {
+                    $nestedCondition = new BasicCondition(...$nestedCondition);
+                }
+            }
+
+            if ($nestedCondition->isEmpty()) {
+                continue;
+            }
+
+            $this->add(clone $nestedCondition);
+        }
+    }
+
+    public function __clone()
+    {
+        foreach ($this->elements as $k => $nestedCondition) {
+            $this->elements[$k] = clone $nestedCondition;
+            $this->elements[$k]->owner = $this;
+        }
+    }
+
+    /**
+     * Return array of nested conditions.
+     *
+     * @return AbstractCondition[]
+     */
+    public function getNestedConditions()
+    {
+        return $this->elements;
+    }
+
+    public function onChangeModel(): void
+    {
+        foreach ($this->elements as $nestedCondition) {
+            $nestedCondition->onChangeModel();
+        }
+    }
+
+    public function isEmpty(): bool
+    {
+        return empty($this->elements);
```

Comment:
count() === 0

Reply:
what is the difference?

Reply:
yes, empty never be used when the input is an array

Reply:
Done